In [1]:
import pandas as pd
import matplotlib.pyplot as plt

Intel MKL WARNING: Support of Intel(R) Streaming SIMD Extensions 4.2 (Intel(R) SSE4.2) enabled only processors has been deprecated. Intel oneAPI Math Kernel Library 2025.0 will require Intel(R) Advanced Vector Extensions (Intel(R) AVX) instructions.
Intel MKL WARNING: Support of Intel(R) Streaming SIMD Extensions 4.2 (Intel(R) SSE4.2) enabled only processors has been deprecated. Intel oneAPI Math Kernel Library 2025.0 will require Intel(R) Advanced Vector Extensions (Intel(R) AVX) instructions.


In [36]:
df = pd.read_csv("data.csv")
display(df.head())

print("Number of columns:", len(df.columns.tolist()))
for col in df.columns.tolist():
    print(col)

incident_id report_id  thread Validated By Title Tags  date_coded  \
0           74         2     NaN          NaN   NaN  NaN  2023-10-29   
1           75         3     NaN          NaN   NaN  NaN  2023-10-29   
2           76         4     NaN          NaN   NaN  NaN  2023-10-29   
3           70         5     NaN          NaN   NaN  NaN  2023-10-23   
4           77         6     NaN          NaN   NaN  NaN  2023-10-29   

                                                 url source file  \
0  https://twitter.com/MikaelThalen/status/150412...    NaN  NaN   
1  https://twitter.com/ThePatriotOasis/status/163...    NaN  NaN   
2  https://twitter.com/SabrinaHalper/status/17136...    NaN  NaN   
3  https://twitter.com/MechaOrvo/status/171479783...    NaN  NaN   
4  https://twitter.com/NotPoliticians/status/1714...    NaN  NaN   

  file_with_watermark screenshot date_posted format  \
0                 NaN        NaN  2022-03-16  video   
1                 NaN        NaN  2023-02-27  video   
2                 NaN        NaN  2023-10-15  video   
3                 NaN        NaN  2023-10-18  video   
4                 NaN        NaN  2023-10-17  video   

                                          transcript  \
0  Ukrainian: шановні захисники бути Президентом ...   
1  The illegal Russian offensive has been swift, ...   
2  My time is coming to an end, and I would like ...   
3  "Israel has hypnotized the world. May Allah aw...   
4  "The very first thing we have to do is raise t...   

                                     summary_content             type  \
0  The deepfake is a video of President Volodymyr...        cheapfake   
1  President Biden is addressing an audience and ...  unclear/unknown   
2  Screenshot of a TikTok video showing Warren Bu...         deepfake   
3  The video shows Donald Trump giving a speech a...         deepfake   
4  The deepfake is a collection of videos showing...         deepfake   

                                        comment_type  \
0  https://www.npr.org/2022/03/16/1087062648/deep...   
1                                                NaN   
2                                                NaN   
3                        Completely fabricated video   
4                                                NaN   

   Needs external review (type) HiveModeration_Score sharer_faketype  \
0                           NaN                  NaN            fake   
1                           NaN                  NaN  real_authentic   
2                           NaN                  NaN            fake   
3                           NaN                  NaN  real_authentic   
4                           NaN                  NaN            fake   

   Needs External Review (sharer_faketype)  external_verification  \
0                                      NaN          verified_fake   
1                                      NaN          verified_fake   
2                                      NaN  authenticity_disputed   
3                                      NaN          verified_fake   
4                                      NaN  authenticity_disputed   

   Needs External Review (external_verification)  \
0                                            NaN   
1                                            NaN   
2                                            NaN   
3                                            NaN   
4                                            NaN   

                                  links_verification  \
0  https://www.npr.org/2022/03/16/1087062648/deep...   
1  The text posted with the deepfake say, "BREAKI...   
2                                                NaN   
3  Comments on the post saying that the post is f...   
4                                                NaN   

                                       evidence_fake watermark  \
0  Poor video quality, mouth doesn't match words ...       yes   
1            Biden does not usually talk that fast.        yes   
2  The mouth movement

Number of columns: 191
incident_id
report_id
thread
Validated By
Title
Tags
date_coded
url
source
file
file_with_watermark
screenshot
date_posted
format
transcript
summary_content
type
comment_type
Needs external review (type)
HiveModeration_Score
sharer_faketype
Needs External Review (sharer_faketype)
external_verification
Needs External Review (external_verification)
links_verification
evidence_fake
watermark
Needs External Verification (watermark)
likes
views
comments
shares
original_source_name
original_source_username
Needs External Review (Original Source Name)
original_source_type
Needs External Review (Original Source Type)
original_source_actor_type
Needs External Review (Original Source Actor Type)
original_source_country
other_original_source_country
Needs External Review (Original Source Country)
comments_coding_original_source
sharer_name
other_sharer_name
Needs External Review (Sharer Name)
sharer_username
sharer_type
Needs External Review (Sharer Type)
sharer_job
Needs E

A lot of columns; check for which ones are redundant (almost all NaN). Find most relevant ones.

In [37]:
all_nan_cols = df.columns[df.isna().all()]
print("Columns with all NaN values:", len(list(all_nan_cols)))

threshold = 0.99
high_nan_cols = df.columns[df.isna().mean() > threshold]
print("Columns with more than 99% NaN values:", len(list(high_nan_cols)))

# Find most relevant columns
df_relevant = df.drop(columns=high_nan_cols)
print("\nRemaining columns:", len(df_relevant.columns.to_list()))
for relev in df_relevant.columns.to_list():
    print(relev)

Columns with all NaN values: 61
Columns with more than 99% NaN values: 112

Remaining columns: 79
incident_id
report_id
Validated By
Title
date_coded
url
file
screenshot
date_posted
format
transcript
summary_content
type
comment_type
HiveModeration_Score
sharer_faketype
external_verification
links_verification
evidence_fake
watermark
likes
views
comments
shares
original_source_name
original_source_type
original_source_actor_type
original_source_country
comments_coding_original_source
sharer_name
other_sharer_name
sharer_username
sharer_type
sharer_job
high_impact_comment
sharer_country
other_sharer_country
sharer_city
other_sharer_city
comments_sharer_coding
target_name
target_one_name
target_one_sentiment
target_one_type_macro
target_one_type_micro
target_one_country
target_two_name
target_two_sentiment
target_two_type_macro
target_two_type_micro
target_two_country
target_three_name
target_three_sentiment
target_three_type_macro
target_three_type_micro
target_three_country
target_type

## Social platforms

In [38]:
from urllib.parse import urlparse

# Determine social platform by extracting the domain from the URL
df['domain'] = df['url'].apply(lambda x: urlparse(x).netloc if pd.notnull(x) else None)
print("Social platform counts:\n", df['domain'].value_counts())

Social platform counts:
 domain
x.com                                     601
twitter.com                               159
www.tiktok.com                             39
youtu.be                                   31
www.youtube.com                            22
www.instagram.com                          18
miro.medium.com                            16
www.facebook.com                           14
farid.berkeley.edu                         12
www.reddit.com                              8
truthsocial.com                             7
archive.is                                  6
                                            4
web.archive.org                             3
www.cnn.com                                 3
www.alternet.org                            3
techcrunch.com                              3
archive.li                                  2
static01.nytimes.com                        2
nypost.com                                  2
archive.ph                                  1
ww

We can see that the most popular deepfake/cheapfake platforms are X/Twitter, TikTok, YouTube, Instagram and Facebook. We have some other websites to explore too, such as Medium, Berkeley and TruthSocial, etc. 

In [39]:
# Count rows per year in the date_posted column
df['date_posted'] = pd.to_datetime(df['date_posted'], errors='coerce')
df['year'] = df['date_posted'].dt.year
print("Rows per year:\n", df['year'].value_counts())

Rows per year:
 year
2024.0    617
2023.0    240
2020.0     52
2021.0     35
2022.0     16
2019.0      9
2018.0      2
2016.0      1
2025.0      1
Name: count, dtype: int64


We see that most posts were made in 2024 and 2023, which can provide a sufficient basis for our analysis (857 rows). 

In [40]:
# Identify rows with media 
media_rows = df[df['file'].notna()]
print("Number of rows with media:", media_rows.shape[0])
print("Proportion with media:", media_rows.shape[0]/len(df))

Number of rows with media: 928
Proportion with media: 0.9508196721311475


Around 95% of the rows have media, which is promising for some image processing analysis. 

In [41]:
print("Format counts:\n", df['format'].value_counts())
print("\nType counts:\n", df['type'].value_counts())
print("\nSharer faketype counts:\n", df['sharer_faketype'].value_counts())
print("\nExternal verification counts:\n", df['external_verification'].value_counts())

Format counts:
 format
image    646
video    330
Name: count, dtype: int64

Type counts:
 type
deepfake           803
unclear/unknown     89
cheapfake           82
real                 1
Name: count, dtype: int64

Sharer faketype counts:
 sharer_faketype
fake              468
unclear           350
real_authentic    157
Name: count, dtype: int64

External verification counts:
 external_verification
unknown                  455
verified_fake            325
authenticity_disputed    156
verified_real             39
Name: count, dtype: int64


The majority of media is images with a significant portion of videos. We have enough on deepfakes to fully focus on it (and not worry about the nuances/differences between deepfakes and cheapfakes, etc). Most of the sharers of these posts deem their posts to be fake, but an equally significant portion does not explicitly state anything (350 sharers being unclear). Huge lack of external verification which can be a potential weakness. 

# Potential ways to filter the data

We will also filter away all the high-NaN columns as well for all options.

In [42]:
df = df.drop(columns=high_nan_cols)

### 1. 2023-2024 analysis of deepfake images/videos that have been verified 

In [43]:
df1 = df[
    (df['year'].isin([2023, 2024])) & 
    (df['type'] == 'deepfake') &       
    (df['file'].notna()) &             
    (df['external_verification'] != 'unknown')
]

In [45]:
print(df1.shape)
display(df1)

(340, 81)


incident_id report_id      Validated By  \
16           203        17               NaN   
18           186        18               NaN   
20           187        19               NaN   
31           201        25               NaN   
32           283        26               NaN   
..           ...       ...               ...   
917         1002       543               NaN   
925         1011       444  Isabella Mulford   
957         1044       594       Alaina Hall   
958         1045       533               NaN   
960         1047       543               NaN   

                                  Title  date_coded  \
16                                  NaN  2024-02-12   
18                                  NaN  2024-02-04   
20                                  NaN  2024-02-05   
31                                  NaN  2024-02-09   
32                                  NaN  2024-03-04   
..                                  ...         ...   
917                                 NaN  2025-03-04   
925  Donald Trump and Elon Musk dancing  2025-03-05   
957                                 NaN  2025-03-05   
958                                 NaN  2025-03-05   
960                                 NaN  2025-03-05   

                                                   url  \
16   https://twitter.com/DouglasLucas/status/171631...   
18   https://twitter.com/21WIRE/status/165300993699...   
20       https://www.facebook.com/reel/894486844951526   
31   https://twitter.com/JackPosobiec/status/163383...   
32   https://www.reddit.com/r/ChatGPT/comments/156h...   
..                                                 ...   
917  https://x.com/cattocrunch/status/1825860652847...   
925  https://x.com/realDonaldTrump/status/182406968...   
957  https://x.com/Jere_Memez/status/18250285930359...   
958  https://x.com/YourMomsHo42215/status/182532763...   
960  https://x.com/cattocrunch/status/1825860652847...   

                                                  file  \
16   Screenshot 2024-02-12 at 12.47.24 PM.png (http...   
18   "Patrick Henningsen on X_ _⭕️ Granted, this mu...   
20   deepfake 19.html (https://v5.airtableuserconte...   
31   Screenshot 2024-02-09 134519.png (https://v5.a...   
32   deepfake 26.html (https://v5.airtableuserconte...   
..                                                 ...   
917  GVQyp6pWAAAk6cw.jpg (https://v5.airtableuserco...   
925  ScreenRecording_03-05-2025 14-04-06_1.mov (htt...   
957  pdid 594.jpg (https://v5.airtableusercontent.c...   
958  pdid 533.jpg (https://v5.airtableusercontent.c...   
960  GVQyp6pWAAAk6cw.jpg (https://v5.airtableuserco...   

                                            screenshot date_posted format  \
16                                                 NaN  2023-10-23  image   
18                                                 NaN  2023-05-01  video   
20                                                 NaN  2023-05-03  video   
31                                                 NaN  2023-03-09  video   
32                                                 NaN  2023-07-22  video   
..                                                 ...         ...    ...   
917  Screenshot 2025-03-03 210934.png (https://v5.a...  2024-08-20  image   
925  Screenshot 2025-03-05 at 2.04.31 PM.png (https...  2024-08-15  video   
957  Screenshot 2024-10-30 at 12.54.40 PM.png (http...  2024-08-18  image   
958  Screen Shot 2025-03-05 at 4.05.44 PM.png (http...  2024-08-18  image   
960  Screenshot 2025-03-03 210934.png (https://v5.a...  2024-08-20  image   

                                            transcript  \
16                                                 NaN   
18   "Today is today, and yesterday was today yeste...   
20   “Today is today, and yesterday is today yester...   
31   I'll tell you something the deepfake that Jack...   
32   "Highly anticipated season for reloaded update...   
..                                                 ...   
917                                       

### 2. 2023-2024 analysis of deepfake images/videos, regardless of verification

In [47]:
df2 = df[
    (df['year'].isin([2023, 2024])) & 
    (df['type'] == 'deepfake') &       
    (df['file'].notna())            
]

In [48]:
print(df2.shape)
display(df2)

(714, 81)


incident_id report_id Validated By Title  date_coded  \
16           203        17          NaN   NaN  2024-02-12   
18           186        18          NaN   NaN  2024-02-04   
20           187        19          NaN   NaN  2024-02-05   
22           188        20          NaN   NaN  2024-02-05   
31           201        25          NaN   NaN  2024-02-09   
..           ...       ...          ...   ...         ...   
971         1058       460          NaN   NaN  2025-03-06   
972         1059       599          NaN   NaN  2025-03-06   
973         1060       757          NaN   NaN  2025-03-06   
974         1061       758          NaN   NaN  2025-03-06   
975         1062       759          NaN   NaN  2025-03-06   

                                                   url  \
16   https://twitter.com/DouglasLucas/status/171631...   
18   https://twitter.com/21WIRE/status/165300993699...   
20       https://www.facebook.com/reel/894486844951526   
22   https://twitter.com/JebraFaushay/status/165937...   
31   https://twitter.com/JackPosobiec/status/163383...   
..                                                 ...   
971  https://x.com/PlanetOfMemes/status/18240700394...   
972  https://x.com/KryptoKarma2024/status/182516139...   
973  https://x.com/Trump_History45/status/183139475...   
974  https://x.com/Trump_History45/status/182900803...   
975  https://x.com/Trump_History45/status/182891374...   

                                                  file  \
16   Screenshot 2024-02-12 at 12.47.24 PM.png (http...   
18   "Patrick Henningsen on X_ _⭕️ Granted, this mu...   
20   deepfake 19.html (https://v5.airtableuserconte...   
22   Deepfake 20.html (https://v5.airtableuserconte...   
31   Screenshot 2024-02-09 134519.png (https://v5.a...   
..                                                 ...   
971  460.png (https://v5.airtableusercontent.com/v3...   
972  599.png (https://v5.airtableusercontent.com/v3...   
973  pdid 757.jpg (https://v5.airtableusercontent.c...   
974  pdid 758.jpg (https://v5.airtableusercontent.c...   
975  pdid 759.jpg (https://v5.airtableusercontent.c...   

                                            screenshot date_posted format  \
16                                                 NaN  2023-10-23  image   
18                                                 NaN  2023-05-01  video   
20                                                 NaN  2023-05-03  video   
22                                                 NaN  2023-05-18  video   
31                                                 NaN  2023-03-09  video   
..                                                 ...         ...    ...   
971  Screenshot 2025-03-05 at 8.45.11 PM.png (https...  2024-08-15  image   
972  Screenshot 2025-03-06 at 3.01.04 PM.png (https...  2024-08-18  image   
973  Screen Shot 2025-03-06 at 3.17.43 PM.png (http...  2024-09-04  image   
974  Screen Shot 2025-03-06 at 3.37.57 PM.png (http...  2024-08-29  image   
975  Screen Shot 2025-03-06 at 3.54.58 PM.png (http...  2024-08-28  image   

                                            transcript  \
16                                                 NaN   
18   "Today is today, and yesterday was today yeste...   
20   “Today is today, and yesterday is today yester...   
22   "Excuse me, excuse me. I need everyone in the ...   
31   I'll tell you something the deepfake that Jack...   
..                                                 ...   
971                                                NaN   
972                                                NaN   
973                                                NaN   
974                                                NaN   
975                                                NaN   

                                       summary_content      type  \
16   Two polaroid photographs, the one one the righ...  deepfake   
18   Kamala Harris is speaking at a rally and her s...  deepfake   
20   The sharer posted a video showing her reacting.